File: dashboard/src/api/interactions.ts
import { getJson } from './client'
import type { InteractionDetailResponse, InteractionListResponse } from './types'

export interface ListInteractionsParams {
  ticket_id?: string
  customer_id?: string
  interaction_type?: string
  has_embedding?: boolean
  limit?: number
  offset?: number
}

export function listInteractions(params: ListInteractionsParams = {}, signal?: AbortSignal) {
  const search = new URLSearchParams()
  if (params.ticket_id) search.set('ticket_id', params.ticket_id)
  if (params.customer_id) search.set('customer_id', params.customer_id)
  if (params.interaction_type) search.set('interaction_type', params.interaction_type)
  if (params.has_embedding !== undefined) search.set('has_embedding', String(params.has_embedding))
  search.set('limit', String(params.limit ?? 50))
  search.set('offset', String(params.offset ?? 0))
  return getJson<InteractionListResponse>(`/api/interactions?${search.toString()}`, signal)
}

export function getInteraction(interactionId: string, topN = 10, signal?: AbortSignal) {
  return getJson<InteractionDetailResponse>(
    `/api/interactions/${interactionId}?top_n=${topN}`,
    signal,
  )
}

File: api/routers/interactions.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from api.deps import get_db
from api.schemas.interactions import InteractionDetailResponse, InteractionListResponse
from api.services import interactions_service
from recommender.models import InteractionType

router = APIRouter(prefix="/api/interactions", tags=["interactions"])


@router.get("", response_model=InteractionListResponse)
def list_interactions(
    ticket_id: uuid.UUID | None = None,
    customer_id: uuid.UUID | None = None,
    interaction_type: InteractionType | None = None,
    has_embedding: bool | None = None,
    limit: int = Query(default=50, le=200),
    offset: int = Query(default=0, ge=0),
    session: Session = Depends(get_db),
) -> InteractionListResponse:
    return interactions_service.list_interactions(
        session, ticket_id, customer_id, interaction_type, has_embedding, limit, offset
    )


@router.get("/{interaction_id}", response_model=InteractionDetailResponse)
def get_interaction(
    interaction_id: uuid.UUID,
    top_n: int = Query(default=10, le=50),
    session: Session = Depends(get_db),
) -> InteractionDetailResponse:
    return interactions_service.get_interaction(session, interaction_id, top_n)
